include/oneapi/dpl/pstl/tuple_impl.h

```diff
@@ -424,16 +424,36 @@ struct tuple<T1, T...>
         return *this;
     }
 
-    friend bool
-    operator==(const tuple& __lhs, const tuple& __rhs)
-    {
-        return __lhs.holder.value == __rhs.holder.value && __lhs.next == __rhs.next;
-    }
-    friend bool
-    operator!=(const tuple& __lhs, const tuple& __rhs)
-    {
-        return !(__lhs == __rhs);
-    }
+// This macro is used to define binary operators for comparison between oneAPI::dpl::__internal::tuple and
+//  std::tuple when they have the same number of template arguments and the individual elements which are invdividually
+//  comparable. These overloads are required because for overload resolution of template arguments, exact match is
+//  required and the compiler does not consider implicit conversions. We also allow std::tuple implementation to
+//  implement the operators themselves, which guarantees a match with oneAPI::dpl::__internal::tuple
+#define _TUPLE_BINARY_OPERATOR_OVERLOAD(__OPERATOR)                                                                    \
+    template <typename... _U, typename = std::enable_if_t<sizeof...(_U) == (sizeof...(T) + 1)>>                        \
+    friend constexpr bool operator __OPERATOR(const tuple& __lhs, const oneapi::dpl::__internal::tuple<_U...>& __rhs)  \
+    {                                                                                                                  \
+        return static_cast<std::tuple<T1, T...>>(__lhs) __OPERATOR static_cast<std::tuple<_U...>>(__rhs);              \
+    }                                                                                                                  \
+    template <typename... _U, typename = std::enable_if_t<sizeof...(_U) == (sizeof...(T) + 1)>>                        \
+    friend constexpr bool operator __OPERATOR(const tuple& __lhs, const std::tuple<_U...>& __rhs)                      \
+    {                                                                                                                  \
+        return static_cast<std::tuple<T1, T...>>(__lhs) __OPERATOR __rhs;                                              \
+    }                                                                                                                  \
+    template <typename... _U, typename = std::enable_if_t<sizeof...(_U) == (sizeof...(T) + 1)>>                        \
+    friend constexpr bool operator __OPERATOR(const std::tuple<_U...>& __lhs, const tuple& __rhs)                      \
+    {                                                                                                                  \
+        return __lhs __OPERATOR static_cast<std::tuple<T1, T...>>(__rhs);                                              \
+    }
```

Comment:
Question for the reviewers:

Does anyone know if there a way to enable forwarding (universal) references and perfect forwarding for something like this without enumerating all combinations of lvalue and rvalue refs for the two input parameters for each function call?

We still need to requires the type is a `std::tuple<U...>` / `oneapi::dpl::__internal::tuple<U...>` respectively, but we would like to use forwarding (universal) references.

Reply:
I've created #1481 as an exploration of another approach.  Please take a look there.

Reply:
closed that other approach in favor of the one here to avoid any copies.